File: tasks/money-rollup/typescript/solution/rollup.ts
const SEGMENT = /^[A-Za-z0-9_]+$/;

export function accountPrefixes(account: unknown): string[] {
  if (typeof account !== "string" || account === "") {
    throw new Error("malformed account");
  }
  const segments = account.split(":");
  if (segments.some((segment) => !SEGMENT.test(segment))) {
    throw new Error("malformed account");
  }
  return segments.map((_, index) => segments.slice(0, index + 1).join(":"));
}

export function rollup(
  items: Array<[unknown, bigint]>
): Array<[string, bigint]> {
  const totals = new Map<string, bigint>();
  for (const [account, minor] of items) {
    let prefixes = accountPrefixes(account);
    if (process.env.LAB_SABOTAGE === "leaf-only-rollup") {
      prefixes = prefixes.slice(-1);
    }
    for (const prefix of prefixes) {
      totals.set(prefix, (totals.get(prefix) ?? 0n) + minor);
    }
  }
  return [...totals.entries()].sort((left, right) =>
    left[0] < right[0] ? -1 : left[0] > right[0] ? 1 : 0
  );
}

File: tasks/money-rollup/go/solution/rollup.go
package main

import (
	"errors"
	"math/big"
	"os"
	"regexp"
	"slices"
	"strings"
)

var segmentPattern = regexp.MustCompile(`^[A-Za-z0-9_]+$`)

type item struct {
	account any
	minor   *big.Int
}

func accountPrefixes(value any) ([]string, error) {
	account, ok := value.(string)
	if !ok || account == "" {
		return nil, errors.New("malformed account")
	}
	segments := strings.Split(account, ":")
	prefixes := make([]string, 0, len(segments))
	for index, segment := range segments {
		if !segmentPattern.MatchString(segment) {
			return nil, errors.New("malformed account")
		}
		prefixes = append(prefixes, strings.Join(segments[:index+1], ":"))
	}
	return prefixes, nil
}

func rollup(items []item) ([]string, map[string]*big.Int, error) {
	totals := map[string]*big.Int{}
	for _, entry := range items {
		prefixes, err := accountPrefixes(entry.account)
		if err != nil {
			return nil, nil, err
		}
		if os.Getenv("LAB_SABOTAGE") == "leaf-only-rollup" {
			prefixes = prefixes[len(prefixes)-1:]
		}
		for _, prefix := range prefixes {
			total, present := totals[prefix]
			if !present {
				total = new(big.Int)
				totals[prefix] = total
			}
			total.Add(total, entry.minor)
		}
	}
	accounts := make([]string, 0, len(totals))
	for account := range totals {
		accounts = append(accounts, account)
	}
	slices.Sort(accounts)
	return accounts, totals, nil
}
